### PR TITLE
Improve date internals

### PR DESCRIPTION
**Pull request by rerobika**
- Optimize year from time calculation
- Force arithmetic operations to int32_t/int64_t whenever possible
- Optimize number conversion in date parse
- Cache local TZA of the date object
- Fix a bug in Date.parse timezone parsing

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik robert.fancsik@h-lab.eu

### CODE REVIEW

**Comment by rerobika** [comment]
Benchmark | Heap (bytes) | Stack (bytes) | Perf (sec) |
----: | ----: | ----: | ----: | 
3d-cube.js | 10992 -> 10992 : 0.000% | 13800 -> 13800 : 0.000% | 1.138 -> 1.138 : -0.037% | 
3d-morph.js | 289232 -> 289232 : 0.000% | 1576 -> 1576 : 0.000% | 1.333 -> 1.325 : +0.642% | 
3d-raytrace.js | 117656 -> 117656 : 0.000% | 2540 -> 2540 : 0.000% | 1.332 -> 1.329 : +0.214% | 
access-binary-trees.js | 27120 -> 27120 : 0.000% | 3432 -> 3432 : 0.000% | 0.777 -> 0.775 : +0.209% | 
access-fannkuch.js | 1792 -> 1792 : 0.000% | 1648 -> 1648 : 0.000% | 2.568 -> 2.570 : -0.097% | 
access-nbody.js | 4600 -> 4600 : 0.000% | 1736 -> 1736 : 0.000% | 1.531 -> 1.526 : +0.290% | 
bitops-3bit-bits-in-byte.js | 1024 -> 1024 : 0.000% | 1592 -> 1592 : 0.000% | 0.716 -> 0.714 : +0.371% | 
bitops-bits-in-byte.js | 984 -> 984 : 0.000% | 1592 -> 1592 : 0.000% | 1.007 -> 1.010 : -0.354% | 
bitops-bitwise-and.js | 752 -> 752 : 0.000% | 1344 -> 1344 : 0.000% | 1.228 -> 1.227 : +0.039% | 
bitops-nsieve-bits.js | 61184 -> 61184 : 0.000% | 1592 -> 1592 : 0.000% | 1.504 -> 1.505 : -0.067% | 
controlflow-recursive.js | 1288 -> 1288 : 0.000% | 61584 -> 61584 : 0.000% | 0.530 -> 0.524 : **+1.144%** | 
crypto-aes.js | 28040 -> 28040 : 0.000% | 2288 -> 2288 : 0.000% | 0.882 -> 0.883 : -0.150% | 
crypto-md5.js | 70968 -> 70968 : 0.000% | 1704 -> 1704 : 0.000% | 0.744 -> 0.747 : -0.376% | 
crypto-sha1.js | 46328 -> 46328 : 0.000% | 1720 -> 1720 : 0.000% | 0.762 -> 0.762 : -0.029% | 
crypto.js | 51704 -> 51704 : 0.000% | 3156 -> 3156 : 0.000% | 8.726 -> 8.721 : +0.056% | 
date-format-tofte.js | 8952 -> 8952 : 0.000% | 2016 -> 2000 : +0.794% | 1.219 -> 1.023 : **+16.046%** | 
date-format-xparb.js | 15144 -> 15144 : 0.000% | 2552 -> 2552 : 0.000% | 1.626 -> 0.704 : **+56.696%** | 
deltablue.js | 455544 -> 455544 : 0.000% | 3540 -> 3540 : 0.000% | 5.437 -> 5.452 : -0.270% | 
math-cordic.js | 2048 -> 2048 : 0.000% | 1592 -> 1592 : 0.000% | 1.714 -> 1.708 : +0.391% | 
math-partial-sums.js | 1512 -> 1512 : 0.000% | 1568 -> 1568 : 0.000% | 1.145 -> 1.157 : -1.074% | 
math-spectral-norm.js | 3264 -> 3264 : 0.000% | 1592 -> 1592 : 0.000% | 0.741 -> 0.740 : +0.183% | 
raytrace.js | 22424 -> 22424 : 0.000% | 5000 -> 5000 : 0.000% | 2.645 -> 2.652 : -0.241% | 
richards.js | 8568 -> 8568 : 0.000% | 2140 -> 2140 : 0.000% | 0.310 -> 0.311 : -0.326% | 
string-base64.js | 90104 -> 90104 : 0.000% | 1640 -> 1640 : 0.000% | 1.421 -> 1.420 : +0.105% | 
string-fasta.js | 4400 -> 4400 : 0.000% | 1576 -> 1576 : 0.000% | 3.451 -> 3.459 : -0.249% | 
Geometric mean: | 0.000% | +0.032% | +3.982% | 

Binary (bytes) | master(a95a0b3539) | patch(95ee731906) | Diff |
----: | ----: | ----: | ----: |
size | 226976 | 226976 | 0 bytes |
.rodata | 20087 | 20111 | +24 bytes |
.dynstr | 446 | 446 | 0 bytes |
.rel.plt | 456 | 456 | 0 bytes |
.interp | 25 | 25 | 0 bytes |
.dynsym | 1008 | 1008 | 0 bytes |
.gnu.hash | 464 | 464 | 0 bytes |
.text | 200072 | 199512 | -560 bytes |
.comment | 85 | 85 | 0 bytes |
.shstrtab | 226 | 226 | 0 bytes |
.data | 8 | 8 | 0 bytes |
.ARM.exidx | 8 | 8 | 0 bytes |
.rel.dyn | 16 | 16 | 0 bytes |
.init | 12 | 12 | 0 bytes |
.got | 244 | 244 | 0 bytes |
.plt | 716 | 716 | 0 bytes |
.note.ABI-tag | 32 | 32 | 0 bytes |
.gnu.version_r | 64 | 64 | 0 bytes |
.bss | 1575096 | 1575096 | 0 bytes |
.fini | 8 | 8 | 0 bytes |
.hash | 408 | 408 | 0 bytes |
.gnu.version | 126 | 126 | 0 bytes |
.fini_array | 4 | 4 | 0 bytes |
.init_array | 4 | 4 | 0 bytes |
.dynamic | 248 | 248 | 0 bytes |
.eh_frame | 4 | 4 | 0 bytes |
.ARM.attributes | 53 | 53 | 0 bytes |

File: jerry-core/ecma/builtin-objects/ecma-builtin-date.c [hunk 1]
```diff
@@ -227,32 +270,29 @@ ecma_date_construct_helper (const ecma_value_t *args, /**< arguments passed to t
...
     }
   }
 
-  ecma_number_t prim_value = ecma_number_make_nan ();
-
   if (!ecma_number_is_nan (date_nums[0]))
   {
     /* 8. */
-    ecma_number_t y = ecma_number_trunc (date_nums[0]);
+    ecma_number_t yi = ecma_number_trunc (date_nums[0]);
```
Comment (ossy-szeged) [inline jerry-core/ecma/builtin-objects/ecma-builtin-date.c:280]: yi is coming from new spec, isn't it? In this case please update references too to the latest spec.

Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-date.c]: Thanks, updated.

**Review by dbatyai — APPROVED** [review]
LGTM

**Review by ossy-szeged — APPROVED** [review]
LGTM

**Pull request merged**